journal/src/main/scala/com/evolutiongaming/kafka/journal/util/ResourcePool.scala

```diff
@@ -0,0 +1,276 @@
+package com.evolutiongaming.kafka.journal.util
+
+import cats.instances.all._
+import cats.effect.implicits._
+import cats.effect.kernel.{Async, Deferred, Sync}
+import cats.effect.std.Mutex
+import cats.effect.{Ref, Resource}
+import cats.syntax.all._
+import com.evolutiongaming.catshelper.Log
+
+import java.time.Instant
+import java.util.UUID
+import java.util.concurrent.TimeoutException
+import scala.concurrent.duration.FiniteDuration
+
+private[journal] trait ResourcePool[F[_], O] {
+
+  /**
+   * `acquire` function of the Resource semantically blocks if there are no objects available.
+   *
+   * `release` function of the Resource returns the object to the pool.
+   * Note that it will NOT deallocate the object.
+   */
+  def borrow: Resource[F, O]
+}
+
+private[journal] object ResourcePool {
+
+  type ObjectId = UUID
+  type WaiterId = UUID
+  type Waiter[F[_], O] = Deferred[F, Either[Throwable, (ObjectId, O)]]
+
+  case class AllocatedObject[F[_], O](value: O, deallocate: F[Unit], state: AllocatedObject.State)
+
+  object AllocatedObject {
+    sealed abstract class State(val isFree: Boolean)
+
+    object State {
+      case object Used extends State(false)
+      case class Free(releasedAt: FiniteDuration) extends State(true)
+    }
+  }
+
+  case class State[F[_], O](
+    objects: Map[ObjectId, AllocatedObject[F, O]],
+    waiters: Map[WaiterId, Waiter[F, O]],
+  )
+
+  def fixedSize[F[_]: Async, O](
+    resource: Resource[F, O],
+    poolSize: Int,
+    acquireTimeout: FiniteDuration,
+    resourceTTL: FiniteDuration,
+    log: Log[F],
+  ): Resource[F, ResourcePool[F, O]] = {
+
+    type StateRef = Ref[F, State[F, O]]
+
+    def poolImpl(mutex: Mutex[F], stateRef: StateRef): ResourcePool[F, O] = {
+
+      def findFreeObject(state: State[F, O]): Option[(ObjectId, AllocatedObject[F, O])] = {
+        state.objects.collectFirst {
+          case (id, obj) if obj.state.isFree => (id, obj)
+        }
+      }
+
+      def markObjectAsUsed(id: ObjectId, obj: AllocatedObject[F, O]): F[Unit] = {
+        stateRef.update { state =>
+          state.copy(objects = state.objects + (id -> obj.copy(state = AllocatedObject.State.Used)))
+        }
+      }
+
+      def allocateObject(state: State[F, O]): F[(ObjectId, O)] = {
+        for {
+          a <- resource.allocated
+          (obj, release) = a
+          objId <- Sync[F].delay(UUID.randomUUID())
+          state1 = state.copy(
+            objects = state.objects + (objId -> AllocatedObject(obj, release, AllocatedObject.State.Used))
+          )
+          _ <- stateRef.set(state1)
+        } yield (objId, obj)
+      }
+
+      def checkForAcquireTimeout(
+        waiterId: WaiterId,
+        promise: Waiter[F, O]
+      ): F[Unit] = {
+        promise.tryGet.flatMap {
+          case Some(_) =>
+            Async[F].unit
+          case None =>
+            mutex.lock.surround {
```

Comment:
I see that `checkForAcquireTimeout` is called from `registerWaiter`, which is called by `acquirePromise`, which already calls `mutex.lock.surround`.

I am actually a bit surprised it does not deadlock in this case as `Mutex` is not reentrant according to documentation.